atomicapp/nulecule/main.py

```diff
@@ -0,0 +1,234 @@
+        self.name = self.namespace = name
+        self.basepath = basepath
+        self.source = source
+        self.params = params or []
+        self.artifacts = artifacts
+        self._app = None
+
+    def load(self):
+        dest = os.path.join(EXTERNAL_APP_DIR, self.name)
+        if not self.artifacts:
+            self.load_external_application(dest)
+
+    def run(self, provider_key, dry=False):
+        if self._app:
+            self._app.run(provider_key, dry)
+            return
+        provider_key, provider = self.get_provider(provider_key, dry)
+        provider.artifacts = self.get_artifact_paths_for_provider(
+            provider_key)
+        provider.init()
+        if not dry:
+            provider.deploy()
+
+    def stop(self, provider_key=None):
+        if self._app:
+            self._app.stop(provider_key)
+        provider_key, provider = self.get_provider(provider_key)
+        provider.init()
+        provider.undeploy()
+
+    def load_config(self, config={}):
+        super(NuleculeComponent, self).load_config(config)
+        if isinstance(self._app, Nulecule):
+            self._app.load_config(config=copy.deepcopy(self.config))
+            self.merge_config(self.config, self._app.config)
+
+    def load_external_application(self, dest):
+        nulecule = Nulecule.unpack(
+            self.source,
+            os.path.join(self.basepath, EXTERNAL_APP_DIR, self.name),
+            namespace=self.namespace)
+        self._app = nulecule
+
+    @property
+    def components(self):
+        if self._app:
+            return self._app.components
+
+    def render(self):
+        if self._app:
+            self._app.render()
+            return
+        context = self.get_context()
+        for provider in self.artifacts:
+            for artifact_path in self.get_artifact_paths_for_provider(
+                    provider):
+                self.render_artifact(artifact_path, context)
+
+    def get_artifact_paths_for_provider(self, provider_key):
+        artifact_paths = []
+        artifacts = self.artifacts.get(provider_key)
+        for artifact in artifacts:
+            if not isinstance(artifact, basestring):
+                continue
+            if artifact.startswith('file:///'):
+                path = artifact[7:]
+            elif artifact.startswith('file://'):
+                path = os.path.join(self.basepath,
+                                    os.path.join(self.basepath, artifact[7:]))
+            else:
+                logger.error('Invalid artifact file')
+                continue
+            artifact_paths.append(path)
+        return artifact_paths
+
+    def render_artifact(self, path, context):
+        with open(path, 'r') as f:
+            content = f.read()
+            template = Template(content)
+            rendered_content = template.safe_substitute(context)
+
+        with open(path, 'w') as f:
+            f.write(rendered_content)
```

Comment:
This is effectively rewriting the original artifact and thus prevents user from modifying `answers.conf` and running again with different params. We are using `.workdir` in cwd currently. That is not correct, obviously, but we need to output the templated artifact somewhere else then to original file.